lib/xilinx/frames.cc

```diff
@@ -35,65 +38,67 @@ int Frames::readFrames(const std::string& frm_file_str) {
 
 		std::vector<std::string> frame_data_strings =
 		    absl::StrSplit(frame_delta.second, ',');
-		if (frame_data_strings.size() != 101) {
-			std::cerr << "Frame " << std::hex << frame_address
-			          << ": found " << std::dec
-			          << frame_data_strings.size()
-			          << "words instead of 101";
-			continue;
-		}
 
-		FrameData frame_data(101, 0);
+		FrameData frame_data(frame_data_strings.size(), 0);
 		std::transform(frame_data_strings.begin(),
 		               frame_data_strings.end(), frame_data.begin(),
 		               [](const std::string& val) -> uint32_t {
 			               return std::stoul(val, nullptr, 16);
 		               });
 
-		// TODO make sure if this is needed
 		updateECC(frame_data);
 
 		// Insert the frame address and corresponding frame data to the
 		// map
-		FrameAddress frm_addr(frame_address);
+		typename ArchType::FrameAddress frm_addr(frame_address);
 		frames_data_.insert(
-		    std::pair<FrameAddress, FrameData>(frm_addr, frame_data));
+		    std::pair<typename ArchType::FrameAddress, FrameData>(
+		        frm_addr, frame_data));
 	}
 	return 0;
 }
 
-void Frames::addMissingFrames(const absl::optional<Part>& part) {
+template <typename ArchType>
+void Frames<ArchType>::addMissingFrames(
+    const absl::optional<typename ArchType::Part>& part) {
 	auto current_frame_address =
-	    absl::optional<FrameAddress>(FrameAddress(0));
+	    absl::optional<typename ArchType::FrameAddress>(
+	        typename ArchType::FrameAddress(0));
 	do {
 		auto iter = frames_data_.find(*current_frame_address);
 		if (iter == frames_data_.end()) {
-			FrameData frame_data(101, 0);
-			// TODO make sure if this is needed
-			updateECC(frame_data);
-			frames_data_.insert(std::pair<FrameAddress, FrameData>(
-			    *current_frame_address, frame_data));
+			FrameData frame_data(ArchType::words_per_frame, 0);
+			frames_data_.insert(
+			    std::pair<typename ArchType::FrameAddress,
+			              FrameData>(*current_frame_address,
+			                         frame_data));
 		}
 		current_frame_address =
 		    part->GetNextFrameAddress(*current_frame_address);
 	} while (current_frame_address);
 }
 
-void Frames::updateECC(FrameData& data) {
+uint32_t calculateECC(const typename Frames<Series7>::FrameData& data) {
```

Comment:
Why did this become a free function?

Reply:
This is a helper function used only by updateECC. I could just have merged it into updateECC, but wanted to follow the single responsibility rule. What I did forget is declaring this function as static, so it won't get linked by anything outside of frames.cc.